target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr941n-v7.c

```diff
@@ -0,0 +1,279 @@
+	unsigned long t;
+
+	base = ioremap(AR71XX_GE0_BASE, AR71XX_GE0_SIZE);
+
+	t = __raw_readl(base + reg);
+
+	iounmap(base);
+
+	return t;
+}
+
+static void __init ath_gmac_reg_wr(unsigned long reg, unsigned long value)
+{
+	void __iomem *base;
+	unsigned long t = value;
+
+	base = ioremap(AR71XX_GE0_BASE, AR71XX_GE0_SIZE);
+
+	__raw_writel(t, base + reg);
+
+	iounmap(base);
+}
+
+static void __init phy_reg_write(unsigned char phy_addr, unsigned char reg, unsigned short data)
+{
+	unsigned short addr = (phy_addr << ATH_ADDR_SHIFT) | reg;
+	volatile int rddata;
+	unsigned short ii = 0xFFFF;
+
+	do
+	{
+		udelay(5);
+		rddata = ath_gmac_reg_rd(ATH_MII_MGMT_IND) & 0x1;
+	} while (rddata && --ii);
+
+	ath_gmac_reg_wr(ATH_MII_MGMT_ADDRESS, addr);
+	ath_gmac_reg_wr(ATH_MII_MGMT_CTRL, data);
+
+	do
+	{
+		udelay(5);
+		rddata = ath_gmac_reg_rd(ATH_MII_MGMT_IND) & 0x1;
+	} while (rddata && --ii);
+}
+
+static unsigned short __init phy_reg_read(unsigned char phy_addr, unsigned char reg)
+{
+	unsigned short addr = (phy_addr << ATH_ADDR_SHIFT) | reg, val;
+	volatile int rddata;
+	unsigned short ii = 0xffff;
+
+	do
+	{
+		udelay(5);
+		rddata = ath_gmac_reg_rd(ATH_MII_MGMT_IND) & 0x1;
+	} while (rddata && --ii);
+
+	ath_gmac_reg_wr(ATH_MII_MGMT_CMD, 0x0);
+	ath_gmac_reg_wr(ATH_MII_MGMT_ADDRESS, addr);
+	ath_gmac_reg_wr(ATH_MII_MGMT_CMD, ATH_MGMT_CMD_READ);
+
+	do
+	{
+		udelay(5);
+		rddata = ath_gmac_reg_rd(ATH_MII_MGMT_IND) & 0x1;
+	} while (rddata && --ii);
+
+	val = ath_gmac_reg_rd(ATH_MII_MGMT_STATUS);
+	ath_gmac_reg_wr(ATH_MII_MGMT_CMD, 0x0);
+
+	return val;
+}
+
+static void __init athrs27_reg_write(unsigned int s27_addr, unsigned int s27_write_data)
+{
+	unsigned int addr_temp;
+	unsigned int data;
+	unsigned char phy_address, reg_address;
+
+	addr_temp = (s27_addr) >> 2;
+	data = addr_temp >> 7;
+
+	phy_address = 0x1f;
+	reg_address = 0x10;
+
+	phy_reg_write(phy_address, reg_address, data);
+
+	phy_address = (0x17 & ((addr_temp >> 4) | 0x10));
+
+	reg_address = (((addr_temp << 1) & 0x1e) | 0x1);
+	data = (s27_write_data >> 16) & 0xffff;
+	phy_reg_write(phy_address, reg_address, data);
+
+	reg_address = ((addr_temp << 1) & 0x1e);
+	data = s27_write_data  & 0xffff;
+	phy_reg_write(phy_address, reg_address, data);
+}
+
+static unsigned int __init athrs27_reg_read(unsigned int s27_addr)
+{
+	unsigned int addr_temp;
+	unsigned int s27_rd_csr_low, s27_rd_csr_high, s27_rd_csr;
+	unsigned int data;
+	unsigned char phy_address, reg_address;
+
+	addr_temp = s27_addr >>2;
+	data = addr_temp >> 7;
+
+	phy_address = 0x1f;
+	reg_address = 0x10;
+
+	phy_reg_write(phy_address, reg_address, data);
+
+	phy_address = (0x17 & ((addr_temp >> 4) | 0x10));
+	reg_address = ((addr_temp << 1) & 0x1e);
+	s27_rd_csr_low = (unsigned int) phy_reg_read(phy_address, reg_address);
+
+	reg_address = reg_address | 0x1;
+	s27_rd_csr_high = (unsigned int) phy_reg_read(phy_address, reg_address);
+	s27_rd_csr = (s27_rd_csr_high << 16) | s27_rd_csr_low ;
+	
+	return (s27_rd_csr);
+}
```

Comment:
Have you noticed that here are some code doing mdio operations  by writing registers directly when you copy code from other people? That's the reason why the original author of this code (hackpascal) thinks that this device support won't be acceptable by openwrt.